File: code_100/java/com/interview/recursion/SetPairTogether.java
package com.interview.recursion;

import java.util.HashMap;
import java.util.Map;

/**
 * Date 12/30/2015
 * @author Tushar Roy
 *
 * Given input array with every element in array having its pair element. How many minimum swaps needs to be
 * done so that all pair elements are adjacent to each other.
 *
 * Time complexity is O(2^n)
 *
 * http://www.geeksforgeeks.org/minimum-number-of-swaps-required-for-arranging-pairs-adjacent-to-each-other/
 */
public class SetPairTogether {

    public int findMinimumSwaps(int input[], Map<Integer, Integer> pair) {
        Map<Integer, Integer> index = new HashMap<>();
        for (int i = 0; i < input.length; i++) {
            index.put(input[i], i);
        }
        return findMinimumSwapsUtil(input, pair, index, 0);
    }

    public int findMinimumSwapsUtil(int input[], Map<Integer, Integer> pair, Map<Integer, Integer> index, int current) {
        if (current == input.length) {
            return 0;
        }

        int v1 = input[current];
        int v2 = input[current + 1];
        int pv2 = pair.get(v1);

        if(pv2 == v2) {
            return findMinimumSwapsUtil(input, pair, index, current + 2);
        } else {
            int idx1 = index.get(v1);
            int idx2 = index.get(v2);

            int idx3 = index.get(pair.get(v1));
            int idx4 = index.get(pair.get(v2));

            swap(index, input, idx2, idx3);
            int val1 = findMinimumSwapsUtil(input, pair, index, current + 2);
            swap(index, input, idx2, idx3);

            swap(index, input, idx1, idx4);
            int val2 = findMinimumSwapsUtil(input, pair, index, current + 2);
            swap(index, input, idx1, idx4);

            return 1 + Math.min(val1, val2);
        }
    }

    private void swap(Map<Integer, Integer> index, int input[], int i, int j) {
        index.compute(input[i], (k, v) -> j);
        index.compute(input[j], (k, v) -> i);

        int t = input[i];
        input[i] = input[j];
        input[j] = t;
    }

    public static void main(String args[]) {
        SetPairTogether spt = new SetPairTogether();
        int input[] = {3, 5, 6, 4, 1, 2};
        Map<Integer, Integer> pair = new HashMap<>();
        pair.put(1, 3);
        pair.put(3, 1);
        pair.put(2, 6);
        pair.put(6, 2);
        pair.put(4, 5);
        pair.put(5 ,4);
        System.out.println(spt.findMinimumSwaps(input, pair));
    }
}

File: code_100/python/recursion/setpairtogether.py
# http://www.geeksforgeeks.org/minimum-number-of-swaps-required-for-arranging-pairs-adjacent-to-each-other/

def find_minimum_swaps(input, pair):
    index = {}
    for i, val in enumerate(input):
        index[val] = i
    return find_minimum_swaps_util(input, pair, index, 0)

def find_minimum_swaps_util(input, pair, index, current):
    if current == len(input):
        return 0

    v1 = input[current]
    v2 = input[current + 1]
    pv2 = pair[v1]

    if pv2 == v2:
        return find_minimum_swaps_util(input, pair, index, current + 2)
    else:
        idx1 = index.get(v1)
        idx2 = index.get(v2)
        idx3 = index.get(pair[v1])
        idx4 = index.get(pair[v2])

        swap(index, input, idx2, idx3)
        val1 = find_minimum_swaps_util(input, pair, index, current+2)
        swap(index, input, idx2, idx3)

        swap(index, input, idx1, idx4)
        val2 = find_minimum_swaps_util(input, pair, index, current+2)
        swap(index, input, idx1, idx4)

        return 1 + max(val1, val2)
        

def swap(index, input, i, j):
    index[input[i]] = j
    index[input[j]] = i
    t = input[i]
    input[i] = input[j]
    input[j] = t

if __name__ == '__main__':
    input = [3, 5, 6, 4, 1, 2]
    pair = {}
    pair[1] = 3
    pair[3] = 1
    pair[2] = 6
    pair[6] = 2
    pair[4] = 5
    pair[5] = 4

    print(find_minimum_swaps(input, pair))
    
